Add keepalived scripts

bzhaoopenstack opened the pull request:
For now, we test with a sample github account, as changing webhook is
very dangerous for online ENV. Also, we post the issue in a sample repo.

Commit:
Add keepalived scripts

For now, we test with a sample github account, as changing webhook is
very dangerous for online ENV.

wangxiyuan commented on etc/keepalived/notify_upstream.py:
```diff
@@ -0,0 +1,34 @@
+    parser.add_argument('--password', metavar='<password>', required=True,
+                        help="The PassWord of Github User")
+    parser.add_argument('--env-role', metavar='<env_role>', required=True,
+                        help="The PassWord of Github User")
```
help message nit.

bzhaoopenstack replied:
Done

wangxiyuan commented on playbooks/keepalived.yaml:
```diff
@@ -47,9 +47,32 @@
+        - { src: "{{ labkeeper_config_git_dest }}/keepalived/master_notify.sh.j2", dest: "/etc/keepalived/master_notify.sh" }
+        - { src: "{{ labkeeper_config_git_dest }}/keepalived/notify_upstream.py", dest: "/etc/keepalived/notify_upstream.py" }
+      vars:
+        upstream_user: moo-ai
```
why moo-ai?

bzhaoopenstack replied:
As we just test , should we add another user "openlab-ci" or "allinone"?  For now, we just test, the formal account should be added at the end of publish the tool.

liusheng replied:
on master host ? once master down...

wangxiyuan replied:
OK, add a todo is better.

bzhaoopenstack replied:
Done

bzhaoopenstack replied:
@liu-sheng yeah, on master host. kp decide which deployment is failed not just based on the host poweroff, also based on the service status. The file name "notify_master" is just used in the case "MASTER ALIVE, BACKUP DOWN".

Yikun commented on etc/keepalived/notify_upstream.py:
```diff
@@ -0,0 +1,34 @@
+    repo = g.get_repo(REPO_NAME)
+    repo.create_issue(
+        title="WARNING: The online openlab deployment <%s> has Down, "
+              "Please recovery asap!" % args.env_role,
```
It's better to add time info in title, let the admins can know the breakdown time.

I prefer something like:

[FATAL][YYYY-MM-DD HH:MM:SS] xxxxxx

bzhaoopenstack replied:
Yeah, it's very similar with your proposal. Like the issue in this repo.;)

Commit:
Fix help message nit

Yikun commented on etc/keepalived/start_slave_svcs.sh.j2:
```diff
@@ -25,3 +25,12 @@ ssh -o StrictHostKeyChecking=no -i /etc/keepalived/openlab.pem "ubuntu@{{ zuul_s
+
+# Get tools to change webhook or change the DNS record(Manually)
+if [ ! -f "/etc/keepalived/cli.py" ];then
+    wget -P /etc/keepalived/  https://raw.githubusercontent.com/wangxiyuan/Toy/master/github-app-webhook-url-update-tool/cli.py
```
If possible we could move god-like-yuan's script into labkeeper, avoid to be removed by accidently.

bzhaoopenstack replied:
god-like-yuan ;-). Make sence to me.

Yikun commented on etc/keepalived/start_slave_svcs.sh.j2:
```diff
@@ -25,3 +25,12 @@ ssh -o StrictHostKeyChecking=no -i /etc/keepalived/openlab.pem "ubuntu@{{ zuul_s
+python3 /etc/keepalived/cli.py --user {{ upstream_user }} --password {{ upstream_password }} --app {{ app_name }} --ip {{ zuul_slave_ip }}
+
+# For now, if a deployment is Down, we need to notify to upstream through a Github Issue
+python /etc/keepalived/notify_upstream.py --username {{ upstream_user }} --password {{ upstream_password }} --env-role {{ ori_env_role }}
```
Any exception or log message should catch in script or in here? at least give user a friendly message, passed or failed.

bzhaoopenstack replied:
This script can not show to user any message, as this script is executed by keepalived, it runs on the backend..So in this script, we must be sure that it can success. And in keepalvied log, we can just see, it runs the script, no more info showed, including the message in the script.

wangxiyuan replied:
This file's name should be updated as well.

bzhaoopenstack replied:
Done

Commit:
Fix yiyun comments

wangxiyuan commented on etc/keepalived/master_notify.sh.j2:
```diff
@@ -0,0 +1,2 @@
+# For now, if a deployment is Down, we need to notify to upstream through a Github Issue
+python /etc/keepalived/change_app_webhook.py --username {{ upstream_user }} --password {{ upstream_password }} --env-role {{ ori_env_role }}
```
I guess this should be `notify_upsteram`

bzhaoopenstack replied:
Done

Commit:
Fix xiyuan comment

huangtianhua commented on playbooks/keepalived.yaml:
```diff
@@ -47,9 +47,35 @@
         keepalived_role: MASTER
         kp_opposite_ip: "{{ hostvars[groups['keepalived-slave'][0]].ansible_host }}"
-        notify_down_action: ""
+        notify_down_action: "/etc/keepalived/master_notify.sh"
```
Use {{ labkeeper_config_git_dest }}/keepalived/master_notify.sh_instead?

huangtianhua replied:
sorry, ignore it.

huangtianhua commented on playbooks/keepalived.yaml:
```diff
@@ -47,9 +47,35 @@
+    - name: Copy necessary scripts
+      copy: src={{ item.src }} dest={{ item.dest }} mode=0755
+      with_items:
+        - { src: "{{ labkeeper_config_git_dest }}/keepalived/master_notify.sh.j2", dest: "/etc/keepalived/master_notify.sh" }
```
same.

liusheng replied:
we cannot use copy to template jinja2 file

huangtianhua commented on playbooks/keepalived.yaml:
```diff
@@ -47,9 +47,35 @@
+      copy: src={{ item.src }} dest={{ item.dest }} mode=0755
+      with_items:
+        - { src: "{{ labkeeper_config_git_dest }}/keepalived/master_notify.sh.j2", dest: "/etc/keepalived/master_notify.sh" }
+        - { src: "{{ labkeeper_config_git_dest }}/keepalived/notify_upstream.py", dest: "/etc/keepalived/notify_upstream.py" }
```
same

huangtianhua commented on playbooks/keepalived.yaml:
```diff
@@ -47,9 +47,35 @@
+        - { src: "{{ labkeeper_config_git_dest }}/keepalived/notify_upstream.py", dest: "/etc/keepalived/notify_upstream.py" }
+      vars:
+        upstream_user: moo-ai
+        upstream_password: !vault |
```
maybe put this to a config file, and then call the config item here?

huangtianhua commented on playbooks/keepalived.yaml:
```diff
@@ -71,16 +97,46 @@
         zuul_master_ip: "{{ hostvars[groups['zuul-scheduler-master'][0]].ansible_host }}"
         zuul_slave_ip: "{{ hostvars[groups['zuul-scheduler-slave'][0]].ansible_host }}"
+        upstream_user: moo-ai
+        upstream_password: !vault |
```
put them to config file.

bzhaoopenstack replied:
Fix then

liusheng commented on playbooks/keepalived.yaml:
```diff
@@ -71,16 +97,46 @@
+    - name: Copy necessary scripts
+      copy: src={{ item.src }} dest={{ item.dest }} mode=0755
+      with_items:
+        - { src: "{{ labkeeper_config_git_dest }}/keepalived/master_notify.sh.j2", dest: "/etc/keepalived/master_notify.sh" }
```
you cannot use copy to replace template

bzhaoopenstack replied:
Fix then

Commit:
Fix comments

liusheng commented on inventory/allinone-ha.yaml:
```diff
@@ -159,9 +159,42 @@ all:
+              61383566333938313831336564623334613362343866383462343461643439333234386532616461
+              3339663232623034660a393662313762366662323536326336373961616664326165363134336134
+              3433
+            ori_env_role: "MASTER"
```
why set `keepalived-slave` to role `MASTER` and `keepalived-master` to role `BACKUP` ?

bzhaoopenstack replied:
These variable just send to the script, in the script we use this variable to let upstream know which deployment is dead. Also the kp deployed toward oppsite depolyment, so let's think, in kp which depolyed in SLAVE deployment, it will monitor whether the MASTER side is DEAD, then the script will post an issue to tell us with a issue topic "The MASTER deployment is DEAD, please XXXX"

liusheng replied:
hmm, got, but it is better to name the variable more clearly, it easy to make confuse

bzhaoopenstack replied:
Fix then

liusheng commented on inventory/allinone-ha.yaml:
```diff
@@ -159,9 +159,42 @@ all:
+            nodepool_master_ip: "{{ hostvars[groups['nodepool-builder-master'][0]].ansible_host }}"
+            nodepool_slave_ip: "{{ hostvars[groups['nodepool-builder-slave'][0]].ansible_host }}"
+            zuul_master_ip: "{{ hostvars[groups['zuul-scheduler-master'][0]].ansible_host }}"
+            zuul_slave_ip: "{{ hostvars[groups['zuul-scheduler-slave'][0]].ansible_host }}"
```
I am not sure if these variables can be the right values if define here, it is better to have a test. and I don't think it is better to move the above 4 variable from playbook to here.

bzhaoopenstack replied:
Already test in allinone with a sample test which is the same with style.. ;-). I think you and Huang core hit a disagreement. Ha. Eventhough I originally think the variable in playbook is much more clear than here.

liusheng replied:
Mainly concern if the above 4 variable can have correct values, but looks like it can. I personally think the above 4 variables should be put in playbook. because these are not configurable values, but following variables are.  the user and password and app_name is related to the real deployment, So Ithink it is ok to put them here.

bzhaoopenstack replied:
Got it, thanks

liusheng commented on playbooks/keepalived.yaml:
```diff
@@ -100,33 +90,12 @@
         - { src: "{{ labkeeper_config_git_dest }}/keepalived/master_notify.sh.j2", dest: "/etc/keepalived/master_notify.sh" }
         - { src: "{{ labkeeper_config_git_dest }}/keepalived/notify_upstream.py", dest: "/etc/keepalived/notify_upstream.py" }
         - { src: "{{ labkeeper_config_git_dest }}/keepalived/slave_notify.sh.j2", dest: "/etc/keepalived/slave_notify.sh" }
         - { src: "{{ labkeeper_config_git_dest }}/keepalived/change_app_webhook.py", dest: "/etc/keepalived/change_app_webhook.py" }
```
can the non-jinja2 template file work ?

bzhaoopenstack replied:
Yeah

liusheng replied:
got, thanks

Commit:
Fix

wangxiyuan commented:
LGTM, don't forget the step to switch DNS A record for status page.

Commit:
Use the same account from all.yaml

But except for kp now, as the script will change the webhook. Use the
test account first before push the public env.
